File: java/amp-bus-rabbit/src/main/java/amp/bus/rabbit/RabbitTransportProvider.java
package amp.bus.rabbit;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.bus.IEnvelopeDispatcher;
import amp.bus.IEnvelopeReceivedCallback;
import amp.bus.ITransportProvider;
import amp.bus.rabbit.topology.ConsumingRoute;
import amp.bus.rabbit.topology.ITopologyService;
import amp.bus.rabbit.topology.ProducingRoute;
import amp.bus.rabbit.topology.RoutingInfo;
import cmf.bus.Envelope;
import cmf.bus.IRegistration;

import com.rabbitmq.client.AMQP.BasicProperties;
import com.rabbitmq.client.Channel;

/**
 * AMQP Implementation of the Transport Provider using the RabbitMQ Client.
 * 
 * @author John Ruiz and Richard Clayton (Berico Technologies)
 */
public class RabbitTransportProvider implements ITransportProvider {

    protected IRabbitChannelFactory channelFactory;
    protected List<IEnvelopeReceivedCallback> envCallbacks = new ArrayList<IEnvelopeReceivedCallback>();
    protected ConcurrentHashMap<IRegistration, RabbitListener> listeners = new ConcurrentHashMap<IRegistration, RabbitListener>();
    protected Logger log;
    protected ITopologyService topologyService;
    
    /**
     * Initialize the Transport Provider with the Topology Service and the Channel Factory.
     * @param topologyService Service that determines the correct exchange and broker to send messages to.
     * @param channelFactory Service that uses topology information to establish connections to the broker.
     */
    public RabbitTransportProvider(ITopologyService topologyService, IRabbitChannelFactory channelFactory) {

		log = LoggerFactory.getLogger(this.getClass());
	
		this.topologyService = topologyService;
		this.channelFactory = channelFactory;
    }

    /**
     * Register a new Envelope handler for the specified routes.
     * @param registration Handlers and routing hints.
     */
    @Override
    public void register(IRegistration registration) throws Exception {
    		
        log.debug("Enter Register");

        // first, get the topology based on the registration info
        RoutingInfo routing = topologyService.getRoutingInfo(registration.getRegistrationInfo());

        for (ConsumingRoute route : routing.getConsumingRoutes()){
        	
        		RabbitListener listener = createListener(registration, route);
        		
        		// store the listener
                listeners.put(registration, listener);
        }

        log.debug("Leave Register");
    }

    protected RabbitListener createListener(
			IRegistration registration, ConsumingRoute route) throws Exception {
    	
    		// create a channel
        Channel channel = channelFactory.getChannelFor(route);

        // create a listener
        RabbitListener listener = this.getListener(registration, route);

        // hook into the listener's events
        listener.onEnvelopeReceived(new IEnvelopeReceivedCallback() {

            @Override
            public void handleReceive(IEnvelopeDispatcher dispatcher) {
            	
                raise_onEnvelopeReceivedEvent(dispatcher);
            }
        });
        
        listener.onClose(new IListenerCloseCallback() {

            @Override
            public void onClose(IRegistration registration) {
            	
                listeners.remove(registration);
            }
        });
        
        listener.onConnectionError(new ReconnectOnConnectionErrorCallback(channelFactory));

        listener.start(channel);
    		
        return listener;
    }
    
    /**
     * Get a new Rabbit Listener for the provided registration and exchange.
     * This was pulled out as an extension point for deriving classes, as well as,
     * to make testing a little easier.
     * 
     * @param registration Handlers and hints
     * @param route Routing Information
     * @return Listener that will pull messages from the broker and call the handlers
     * on the registration.
     */
    protected RabbitListener getListener(IRegistration registration, ConsumingRoute route){
    		
    		return new RabbitListener(registration, route);
    }
    
    @Override
    public void send(Envelope env) throws Exception {
    	
        log.debug("Enter Send");

        // first, get the topology based on the headers
        RoutingInfo routing = topologyService.getRoutingInfo(env.getHeaders());

        for (ProducingRoute route : routing.getProducingRoutes()){
        	
        		log.info("Sending to exchange: " + route.getExchange().getName());
        		
        		Channel channel = null;
        		
        		try {
        			
        			channel = channelFactory.getChannelFor(route);
        			
        			BasicProperties props = 
        				new BasicProperties.Builder()
        					.headers(
        						new HashMap<String, Object>(env.getHeaders()))
        					.build();
        			
        			if (route.getExchange().shouldDeclare()){
        				
        				channel.exchangeDeclare(
        					route.getExchange().getName(), route.getExchange().getExchangeType(), 
        					route.getExchange().isDurable(), route.getExchange().isAutoDelete(), 
        					route.getExchange().getArguments());
        			}
        			
        			for (String routingKey : route.getRoutingKeys()){
        				
        				channel.basicPublish(route.getExchange().getName(), routingKey, props, env.getPayload());
        			}
        			
        		} catch (Exception e) {
        			
             	log.error("Failed to send an envelope", e);
             	throw e;
             } 
        }
        
        log.debug("Leave Send");
    }

    /**
     * Unregister a registration with the bus (canceling the listener
     * and stopping consumption from the broker).
     * @param registration The original registration used to create
     * the listener.
     */
    @Override
    public void unregister(IRegistration registration) {
    	
    		RabbitListener listener = listeners.remove(registration);
    	
        if (listener != null) {
            
        		listener.stopListening();
        }
    }
    
    /**
     * Register a callback for the EnvelopeReceived event.
     * @param callback Called when an envelope is received.
     */
    @Override
    public void onEnvelopeReceived(IEnvelopeReceivedCallback callback) {
        envCallbacks.add(callback);
    }

    /**
     * Used to signal EnvelopeReceived listeners that a new envelope has been
     * received from a listener.
     * @param dispatcher Object used to grab the envelope as well as to complete
     * the process of receiving the message.
     */
    protected void raise_onEnvelopeReceivedEvent(IEnvelopeDispatcher dispatcher) {
        for (IEnvelopeReceivedCallback callback : envCallbacks) {
            try {
                callback.handleReceive(dispatcher);
            } catch (Exception ex) {
                log.error("Caught an unhandled exception raising the onEnvelopeReceived event", ex);
            }
        }
    }
    
    /**
     * Gracefully shutdown all the services associated with the Transport Provider.
     */
    @Override
    public void dispose() {
    		
        try {  channelFactory.dispose(); } catch (Exception ex) { }

        try {  topologyService.dispose(); } catch (Exception ex) { }

        for (RabbitListener l : listeners.values()) {
            
        		try { l.dispose(); } catch (Exception ex) { }
        }
    }

    /**
     * Called before the instance is recycled.
     */
    @Override
    protected void finalize() {
        dispose();
    }
}

File: dotnet/source/amp.bus.rabbit/RabbitTransportProvider.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

using Common.Logging;
using RabbitMQ.Client;

using cmf.bus;
using amp.bus;
using amp.rabbit;
using amp.rabbit.topology;

namespace amp.bus.rabbit
{
    public class RabbitTransportProvider : ITransportProvider, IDisposable
    {
        public event Action<IEnvelopeDispatcher> OnEnvelopeReceived;


        protected IDictionary<IRegistration, RabbitListener> _listeners;
        protected ITopologyService _topoSvc;
        protected IRabbitConnectionFactory _connFactory;
        protected ILog _log;


        public RabbitTransportProvider(
            ITopologyService topologyService, 
            IRabbitConnectionFactory connFactory)
        {
            _topoSvc = topologyService;
            _connFactory = connFactory;

            _listeners = new Dictionary<IRegistration, RabbitListener>();

            _log = LogManager.GetLogger(this.GetType());
        }


        public void Send(Envelope env)
        {
            _log.Debug("Enter Send");

            // first, get the topology based on the headers
            RoutingInfo routing = _topoSvc.GetRoutingInfo(env.Headers);

            // next, pull out all the producer exchanges
            IEnumerable<Exchange> exchanges =
                from route in routing.Routes
                select route.ProducerExchange;

            // for each exchange, send the envelope
            foreach (Exchange ex in exchanges)
            {
                _log.Debug("Sending to exchange: " + ex.ToString());
                IConnection conn = _connFactory.ConnectTo(ex);
                
                using (IModel channel = conn.CreateModel())
                {
                    IBasicProperties props = channel.CreateBasicProperties();
                    props.Headers = env.Headers as IDictionary;

                    channel.ExchangeDeclare(ex.Name, ex.ExchangeType, ex.IsDurable, ex.IsAutoDelete, ex.Arguments);
                    channel.BasicPublish(ex.Name, ex.RoutingKey, props, env.Payload);

                    // close the channel, but not the connection.  Channels are cheap.
                    channel.Close();
                }
            }

            _log.Debug("Leave Send");
        }

        public void Register(IRegistration registration)
        {
            _log.Debug("Enter Register");

            // first, get the topology based on the registration info
            RoutingInfo routing = _topoSvc.GetRoutingInfo(registration.Info);

            // next, pull out all the consumer exchanges
            IEnumerable<Exchange> exchanges =
                from route in routing.Routes
                select route.ConsumerExchange;

            foreach (Exchange ex in exchanges)
            {
                IConnection conn = _connFactory.ConnectTo(ex);

                // create a listener
                RabbitListener listener = new RabbitListener(registration, ex, conn);
                listener.OnEnvelopeReceived += this.listener_OnEnvelopeReceived;
                listener.OnClose += _listeners.Remove;

                // put it on another thread so as not to block this one
                Thread listenerThread = new Thread(listener.Start);
                listenerThread.Name = string.Format("{0} on {1}:{2}{3}", ex.QueueName, ex.HostName, ex.Port, ex.VirtualHost);
                listenerThread.Start();

                // store the listener
                _listeners.Add(registration, listener);
            }

            _log.Debug("Leave Register");
        }

        public virtual void Unregister(IRegistration registration)
        {
            if (_listeners.ContainsKey(registration))
            {
                RabbitListener listener = _listeners[registration];
                listener.Stop();

                _listeners.Remove(registration);
            }
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }


        protected virtual void listener_OnEnvelopeReceived(IEnvelopeDispatcher dispatcher)
        {
            if (null != this.OnEnvelopeReceived)
            {
                foreach (Delegate d in this.OnEnvelopeReceived.GetInvocationList())
                {
                    try { d.DynamicInvoke(dispatcher); }
                    catch { }
                }
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            _log.Debug("Enter Dispose");

            if (disposing)
            {
                // get rid of managed resources
                try { _listeners.Values.ToList().ForEach(l => l.Stop()); }
                catch { }

                try { _connFactory.Dispose(); }
                catch { }
            }
            // get rid of unmanaged resources

            _log.Debug("Leave Dispose");
        }
    }
}
